File: app/services/opencodeReasoningSettle.test.ts
import { describe, expect, test } from "bun:test";

import { normalizeCodexConversation } from "./codexConversation";
import { projectZenTimeline } from "../components/terminal/projectZenTimeline";
import {
  reconcileConversationDeltaEvents,
  reconcileConversationSnapshot,
} from "../components/terminal/interfaceConversationReconciliation";
import type { ZenTimelineItem } from "../components/terminal/InterfaceTimelineItemView";

const T = (sec: number) =>
  new Date(Date.parse("2026-08-08T00:00:10Z") + sec * 1000).toISOString();

/**
 * Daemon wire shape after the parser settle fix. OpenCode's real terminal
 * message carries finish "unknown" + time.completed; the parser settles the
 * turn and clears Partial on every event (Go omitempty), so a completed
 * reasoning part arrives with partial absent. A live turn keeps partial on the
 * reasoning part until the authoritative completion boundary lands.
 */
interface OpenCodeReasoningDaemonEvent {
  id: string;
  seq: number;
  kind: string;
  role?: "user" | "assistant";
  body?: string;
  tool_name?: string;
  call_id?: string;
  input?: string;
  output?: string;
  status?: string;
  partial?: boolean;
  transient?: boolean;
  timestamp?: string;
}

function reasoningTurnDaemonEvents(live: boolean): OpenCodeReasoningDaemonEvent[] {
  const reasoning = {
    id: "prt_reason_final",
    seq: 4,
    kind: "reasoning",
    body: "Now I understand the scheduler",
    timestamp: T(4),
    transient: true,
  };
  return [
    {
      id: "prt_user",
      seq: 1,
      kind: "user_message",
      role: "user",
      body: "fix the scheduler",
      timestamp: T(1),
    },
    {
      id: "prt_reason_step1",
      seq: 2,
      kind: "reasoning",
      body: "first step thinking",
      timestamp: T(2),
      transient: true,
      partial: live ? true : undefined,
    },
    {
      id: "prt_tool",
      seq: 3,
      kind: "tool_call",
      tool_name: "bash",
      call_id: "call_1",
      input: '{"command":"go build"}',
      output: live ? undefined : "ok",
      status: live ? "running" : "completed",
      partial: live ? true : undefined,
      timestamp: T(3),
    },
    {
      ...reasoning,
      partial: live ? true : undefined,
    },
    {
      id: "prt_text",
      seq: 5,
      kind: "assistant_message",
      role: "assistant",
      body: "done",
      timestamp: T(5),
      partial: live ? true : undefined,
    },
  ];
}

function reasoningDaemonConversation(events: OpenCodeReasoningDaemonEvent[]) {
  const live = events.some(
    (event) => event.partial === true || event.status === "running",
  );
  return {
    available: true,
    session_id: "ses_rsn",
    activity: {
      id: "ses_rsn:activity:prt_user",
      status: live ? "running" : "completed",
      started_at: T(1),
      settled_at: live ? undefined : T(6),
    },
    events,
  };
}

function reasoningItem(
  items: ReturnType<typeof projectZenTimeline>["items"],
  id = "prt_reason_final",
) {
  return items.find(
    (item): item is Extract<ZenTimelineItem, { type: "activity" }> =>
      item.id === id && item.type === "activity",
  );
}

describe("OpenCode reasoning icon settle contract", () => {
  test("completed historical reasoning renders settled static state with loaded text", () => {
    const conversation = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(false)),
    );
    const { items } = projectZenTimeline(conversation.events, null);
    const reason = reasoningItem(items);
    expect(reason).toMatchObject({
      type: "activity",
      activityKind: "reasoning",
      tone: "neutral",
      icon: "bulb",
      body: "Now I understand the scheduler",
      defaultExpanded: false,
    });
    // Completed history must never keep the spinner source alive. The daemon
    // omits partial on settled rows (Go omitempty), so streaming is falsy.
    expect(reason?.streaming).toBeFalsy();
    expect(reason).not.toMatchObject({ tone: "running" });
  });

  test("active stream keeps the spinner only on the live reasoning row", () => {
    const conversation = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(true)),
    );
    const { items } = projectZenTimeline(conversation.events, null);
    expect(reasoningItem(items, "prt_reason_step1")).toMatchObject({
      tone: "running",
      streaming: true,
      defaultExpanded: true,
    });
    expect(reasoningItem(items)).toMatchObject({
      tone: "running",
      streaming: true,
      body: "Now I understand the scheduler",
    });
  });

  test("incremental terminal delta settles the exact reasoning row without duplicates", () => {
    const live = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(true)),
    );
    const settled = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(false)),
    );
    const deltaUpserts = settled.events.filter(
      (event) =>
        event.partial !==
        live.events.find((candidate) => candidate.id === event.id)?.partial ||
        event.status !==
        live.events.find((candidate) => candidate.id === event.id)?.status,
    );
    const reconciled = reconcileConversationDeltaEvents(
      live.events,
      deltaUpserts,
    );
    const byId = new Map(reconciled.map((event) => [event.id, event]));
    expect(byId.size).toBe(reconciled.length);
    expect(byId.size).toBe(5);
    const { items } = projectZenTimeline(reconciled, null);
    expect(reasoningItem(items, "prt_reason_step1")).toMatchObject({
      tone: "neutral",
    });
    expect(reasoningItem(items, "prt_reason_step1")?.streaming).toBeFalsy();
    expect(reasoningItem(items)).toMatchObject({
      tone: "neutral",
      body: "Now I understand the scheduler",
    });
    expect(reasoningItem(items)?.streaming).toBeFalsy();
    const reasoningRows = items.filter(
      (item) => item.type === "activity" && item.activityKind === "reasoning",
    );
    expect(reasoningRows).toHaveLength(2);
  });

  test("reconnect/cache revisit snapshot replacement settles completed history", () => {
    const live = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(true)),
    );
    const settled = normalizeCodexConversation(
      reasoningDaemonConversation(reasoningTurnDaemonEvents(false)),
    );
    const revisited = reconcileConversationSnapshot(live, settled, true);
    const { items } = projectZenTimeline(revisited.events, null);
    const ids = items.map((item) => item.id);
    expect(new Set(ids).size).toBe(ids.length);
    expect(reasoningItem(items)).toMatchObject({
      tone: "neutral",
      body: "Now I understand the scheduler",
    });
    expect(reasoningItem(items)?.streaming).toBeFalsy();
  });

  test("tool interleaving stays correct on both live and settled projections", () => {
    const liveItems = projectZenTimeline(
      normalizeCodexConversation(
        reasoningDaemonConversation(reasoningTurnDaemonEvents(true)),
      ).events,
      null,
    ).items;
    expect(reasoningItem(liveItems, "prt_tool")).toMatchObject({
      tone: "running",
    });
    expect(reasoningItem(liveItems, "prt_tool")?.statusKey).toBe("running");

    const settledItems = projectZenTimeline(
      normalizeCodexConversation(
        reasoningDaemonConversation(reasoningTurnDaemonEvents(false)),
      ).events,
      null,
    ).items;
    expect(reasoningItem(settledItems, "prt_tool")).toMatchObject({
      tone: "success",
      statusKey: "completed",
    });
    expect(reasoningItem(settledItems)).toMatchObject({ tone: "neutral" });
  });

  test("cancelled/interrupted convergence settles while unresolved live parts stay running", () => {
    const interrupted = normalizeCodexConversation({
      ...reasoningDaemonConversation(reasoningTurnDaemonEvents(false)),
      activity: {
        id: "ses_rsn:activity:prt_user",
        status: "interrupted",
        started_at: T(1),
        settled_at: T(6),
      },
    });
    const interruptedItems = projectZenTimeline(interrupted.events, null).items;
    expect(reasoningItem(interruptedItems)).toMatchObject({
      tone: "neutral",
    });
    expect(reasoningItem(interruptedItems)?.streaming).toBeFalsy();

    // A genuinely unresolved live part (no terminal boundary) must keep the
    // spinner: fail-closed, never settled by text availability alone.
    const unresolved = normalizeCodexConversation(
      reasoningDaemonConversation([
        ...reasoningTurnDaemonEvents(true),
        {
          id: "prt_reason_open",
          seq: 6,
          kind: "reasoning",
          body: "still streaming thoughts",
          timestamp: T(7),
          transient: true,
          partial: true,
        },
      ]),
    );
    const unresolvedItems = projectZenTimeline(unresolved.events, null).items;
    expect(reasoningItem(unresolvedItems, "prt_reason_open")).toMatchObject({
      tone: "running",
      streaming: true,
      body: "still streaming thoughts",
    });
  });
});

File: daemon/work/opencode_reasoning_settle_test.go
package work

import (
	"fmt"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/daoleno/zen/daemon/classifier"
)

// openCodeReasoningTurnSeeds builds the real OpenCode v1.18 terminal-turn
// shape observed in production: the last assistant message carries
// finish:"unknown" with the authoritative time.completed marker, its parts are
// step-start / reasoning / step-finish(reason:"unknown"), and every earlier
// step closed with tool-calls. Prior parser logic treated "unknown" like a
// live tool-calls yield, so the turn never settled and the reasoning event
// stayed partial — the Interface reasoning icon spun forever on completed
// history.
func openCodeReasoningTurnSeeds(started time.Time, terminalMessageData string) ([]openCodeMessageSeed, []openCodePartSeed) {
	messages := []openCodeMessageSeed{
		{ID: "msg_user", SessionID: "ses_rsn", CreatedMS: started.Add(time.Second).UnixMilli(), Data: `{"role":"user","time":{"created":1}}`},
		{ID: "msg_yield", SessionID: "ses_rsn", CreatedMS: started.Add(2 * time.Second).UnixMilli(), Data: fmt.Sprintf(`{"role":"assistant","finish":"tool-calls","time":{"created":1,"completed":%d}}`, started.Add(5*time.Second).UnixMilli())},
		{ID: "msg_terminal", SessionID: "ses_rsn", CreatedMS: started.Add(10 * time.Second).UnixMilli(), Data: terminalMessageData},
	}
	parts := []openCodePartSeed{
		{ID: "p_user", MessageID: "msg_user", SessionID: "ses_rsn", CreatedMS: started.Add(time.Second).UnixMilli(), Data: `{"type":"text","text":"task"}`},
		{ID: "p_s1", MessageID: "msg_yield", SessionID: "ses_rsn", CreatedMS: started.Add(2 * time.Second).UnixMilli(), Data: `{"type":"step-start"}`},
		{ID: "p_x1", MessageID: "msg_yield", SessionID: "ses_rsn", CreatedMS: started.Add(3 * time.Second).UnixMilli(), Data: `{"type":"reasoning","text":"first step thinking"}`},
		{ID: "p_t1", MessageID: "msg_yield", SessionID: "ses_rsn", CreatedMS: started.Add(4 * time.Second).UnixMilli(), Data: `{"type":"tool","tool":"bash","callID":"call-1","state":{"status":"completed","input":{"command":"true"},"output":"ok"}}`},
		{ID: "p_f1", MessageID: "msg_yield", SessionID: "ses_rsn", CreatedMS: started.Add(5 * time.Second).UnixMilli(), Data: `{"type":"step-finish","reason":"tool-calls"}`},
		{ID: "p_s2", MessageID: "msg_terminal", SessionID: "ses_rsn", CreatedMS: started.Add(10 * time.Second).UnixMilli(), Data: `{"type":"step-start"}`},
		{ID: "p_x2", MessageID: "msg_terminal", SessionID: "ses_rsn", CreatedMS: started.Add(11 * time.Second).UnixMilli(), Data: `{"type":"reasoning","text":"terminal reasoning loaded"}`},
		{ID: "p_f2", MessageID: "msg_terminal", SessionID: "ses_rsn", CreatedMS: started.Add(14 * time.Second).UnixMilli(), Data: `{"type":"step-finish","reason":"unknown"}`},
	}
	return messages, parts
}

// TestOpenCodeUnknownFinishedTerminalMessageSettlesCompletedTurn is the red
// regression for the spinner defect: OpenCode's real terminal message carries
// finish:"unknown" + time.completed, and the turn must settle completed with
// the reasoning event losing Partial so the Interface settles the icon.
func TestOpenCodeUnknownFinishedTerminalMessageSettlesCompletedTurn(t *testing.T) {
	started := time.Date(2026, 8, 8, 0, 0, 10, 0, time.UTC)
	terminalData := fmt.Sprintf(
		`{"role":"assistant","finish":"unknown","time":{"created":1,"completed":%d}}`,
		started.Add(14*time.Second).UnixMilli(),
	)
	messages, parts := openCodeReasoningTurnSeeds(started, terminalData)
	dbPath := filepath.Join(t.TempDir(), "opencode.db")
	createOpenCodeFixtureDB(t, dbPath, []openCodeSessionSeed{
		{ID: "ses_rsn", Directory: "/repo", CreatedMS: started.UnixMilli(), UpdatedMS: started.Add(15 * time.Second).UnixMilli()},
	}, messages, parts)
	got, err := parseOpenCodeConversation(dbPath, "ses_rsn")
	if err != nil {
		t.Fatal(err)
	}
	if got.Activity == nil || got.Activity.Status != ProviderActivityCompleted {
		t.Fatalf("unknown+completed terminal message must settle completed: %+v", got.Activity)
	}
	wantSettled := normalizeCodexTimestamp(started.Add(14 * time.Second).UTC().Format(time.RFC3339Nano))
	if got.Activity.SettledAt != wantSettled {
		t.Fatalf("settled at = %q, want terminal time.completed %q", got.Activity.SettledAt, wantSettled)
	}
	var reasoning *CodexConversationEvent
	for i := range got.Events {
		if got.Events[i].Kind == "reasoning" && strings.Contains(got.Events[i].Body, "terminal reasoning") {
			reasoning = &got.Events[i]
		}
	}
	if reasoning == nil {
		t.Fatalf("terminal reasoning event missing: %#v", got.Events)
	}
	if reasoning.Partial {
		t.Fatalf("settled reasoning must clear Partial (spinner source): %#v", *reasoning)
	}
	if !strings.Contains(reasoning.Body, "terminal reasoning loaded") {
		t.Fatalf("loaded reasoning text must be preserved: %q", reasoning.Body)
	}
}

// TestOpenCodeUnknownFinishWithoutCompletedStaysLive keeps the fail-closed
// guarantee: a message row still mid-write (no time.completed) must never
// settle, and the reasoning part stays genuinely live.
func TestOpenCodeUnknownFinishWithoutCompletedStaysLive(t *testing.T) {
	started := time.Date(2026, 8, 8, 0, 0, 10, 0, time.UTC)
	messages, parts := openCodeReasoningTurnSeeds(started, `{"role":"assistant","finish":"unknown","time":{"created":1}}`)
	dbPath := filepath.Join(t.TempDir(), "opencode.db")
	createOpenCodeFixtureDB(t, dbPath, []openCodeSessionSeed{
		{ID: "ses_rsn", Directory: "/repo", CreatedMS: started.UnixMilli(), UpdatedMS: started.Add(15 * time.Second).UnixMilli()},
	}, messages, parts)
	got, err := parseOpenCodeConversation(dbPath, "ses_rsn")
	if err != nil {
		t.Fatal(err)
	}
	if got.Activity == nil || got.Activity.Status != ProviderActivityRunning {
		t.Fatalf("unknown finish without completed must stay live: %+v", got.Activity)
	}
	for _, event := range got.Events {
		if event.Kind == "reasoning" && strings.Contains(event.Body, "terminal reasoning") && !event.Partial {
			t.Fatalf("live reasoning must stay partial: %#v", event)
		}
	}
}

// TestOpenCodeUnknownFinishWithRunningToolStaysLive verifies the terminal
// boundary is refused while a tool call is still in flight.
func TestOpenCodeUnknownFinishWithRunningToolStaysLive(t *testing.T) {
	started := time.Date(2026, 8, 8, 0, 0, 10, 0, time.UTC)
	terminalData := fmt.Sprintf(
		`{"role":"assistant","finish":"unknown","time":{"created":1,"completed":%d}}`,
		started.Add(14*time.Second).UnixMilli(),
	)
	messages, parts := openCodeReasoningTurnSeeds(started, terminalData)
	parts = append(parts, openCodePartSeed{
		ID: "p_t2", MessageID: "msg_terminal", SessionID: "ses_rsn",
		CreatedMS: started.Add(12 * time.Second).UnixMilli(),
		Data:      `{"type":"tool","tool":"bash","callID":"call-2","state":{"status":"running","input":{"command":"sleep"}}}`,
	})
	dbPath := filepath.Join(t.TempDir(), "opencode.db")
	createOpenCodeFixtureDB(t, dbPath, []openCodeSessionSeed{
		{ID: "ses_rsn", Directory: "/repo", CreatedMS: started.UnixMilli(), UpdatedMS: started.Add(15 * time.Second).UnixMilli()},
	}, messages, parts)
	got, err := parseOpenCodeConversation(dbPath, "ses_rsn")
	if err != nil {
		t.Fatal(err)
	}
	if got.Activity == nil || got.Activity.Status != ProviderActivityRunning {
		t.Fatalf("unknown+completed with a running tool must stay live: %+v", got.Activity)
	}
}

// TestOpenCodeTerminalDeltaConvergesReasoningPartial exercises the
// incremental cache + changed-id delta owner: poll 1 sees the live turn
// (reasoning partial), poll 2 lands only the terminal message row update
// (finish unknown + time.completed), and the second load must report the
// reasoning event id in ChangedEventIDs so the server's memoized delta
// carries the Partial flip to the App.
func TestOpenCodeTerminalDeltaConvergesReasoningPartial(t *testing.T) {
	started := time.Date(2026, 8, 8, 0, 0, 10, 0, time.UTC)
	session := []openCodeSessionSeed{
		{ID: "ses_rsn", Directory: "/repo", CreatedMS: started.UnixMilli(), UpdatedMS: started.Add(15 * time.Second).UnixMilli()},
	}
	liveMessages, parts := openCodeReasoningTurnSeeds(started, `{"role":"assistant","finish":"unknown","time":{"created":1}}`)
	dbPath := filepath.Join(t.TempDir(), "opencode.db")
	createOpenCodeFixtureDB(t, dbPath, session, liveMessages, parts)
	t.Setenv("ZEN_OPENCODE_DB", dbPath)
	reader := NewProviderConversationReader()
	agent := classifier.Agent{Cwd: "/repo", Command: "opencode", StartedAt: started}
	first, err := reader.Load(agent, AgentProviderOpenCode, started.Add(time.Minute))
	if err != nil {
		t.Fatal(err)
	}
	var reasoningID string
	for _, event := range first.Events {
		if event.Kind == "reasoning" && strings.Contains(event.Body, "terminal reasoning") {
			reasoningID = event.ID
			if !event.Partial {
				t.Fatalf("live reasoning must be partial: %#v", event)
			}
		}
	}
	if reasoningID == "" {
		t.Fatal("reasoning event missing from live poll")
	}
	version := reader.ConversationVersion()
	if version == 0 {
		t.Fatal("no content version")
	}
	terminalMessages := []openCodeMessageSeed{
		liveMessages[0], liveMessages[1],
		{ID: "msg_terminal", SessionID: "ses_rsn", CreatedMS: started.Add(10 * time.Second).UnixMilli(), Data: fmt.Sprintf(`{"role":"assistant","finish":"unknown","time":{"created":1,"completed":%d}}`, started.Add(14*time.Second).UnixMilli())},
	}
	createOpenCodeFixtureDB(t, dbPath, session, terminalMessages, parts)
	second, err := reader.Load(agent, AgentProviderOpenCode, started.Add(time.Minute))
	if err != nil {
		t.Fatal(err)
	}
	if reader.ConversationVersion() == version {
		t.Fatal("terminal settle must bump the content version")
	}
	if second.Activity == nil || second.Activity.Status != ProviderActivityCompleted {
		t.Fatalf("terminal delta must settle: %+v", second.Activity)
	}
	for _, event := range second.Events {
		if event.ID == reasoningID && event.Partial {
			t.Fatalf("delta-converged reasoning must clear Partial: %#v", event)
		}
	}
	changed := map[string]bool{}
	for _, id := range reader.ChangedEventIDs() {
		changed[id] = true
	}
	if !changed[reasoningID] {
		t.Fatalf("settled Partial flip not reported as changed for memoized delta: %q", reasoningID)
	}
}

// TestOpenCodeInterruptedTerminalMessageConverges covers cancellation: an
// aborted/failed finish settles the turn interrupted/failed and clears the
// reasoning Partial.
func TestOpenCodeInterruptedTerminalMessageConverges(t *testing.T) {
	started := time.Date(2026, 8, 8, 0, 0, 10, 0, time.UTC)
	terminalData := fmt.Sprintf(
		`{"role":"assistant","finish":"aborted","time":{"created":1,"completed":%d}}`,
		started.Add(14*time.Second).UnixMilli(),
	)
	messages, parts := openCodeReasoningTurnSeeds(started, terminalData)
	dbPath := filepath.Join(t.TempDir(), "opencode.db")
	createOpenCodeFixtureDB(t, dbPath, []openCodeSessionSeed{
		{ID: "ses_rsn", Directory: "/repo", CreatedMS: started.UnixMilli(), UpdatedMS: started.Add(15 * time.Second).UnixMilli()},
	}, messages, parts)
	got, err := parseOpenCodeConversation(dbPath, "ses_rsn")
	if err != nil {
		t.Fatal(err)
	}
	if got.Activity == nil || got.Activity.Status != ProviderActivityInterrupted {
		t.Fatalf("aborted terminal finish must settle interrupted: %+v", got.Activity)
	}
	for _, event := range got.Events {
		if event.Kind == "reasoning" && event.Partial {
			t.Fatalf("interrupted reasoning must clear Partial: %#v", event)
		}
	}
}
